### DISCUSSION TITLE
How to score our models ?

### DISCUSSION THREAD

**Discussion by MarcBresson** [Cranfield-GDP3/TACTUS-model#6] [Discussion]
# Model methods

All the models we are using have:
- a `fit(X, y)` method that returns the fitted classifier.
- a `predict(X)` method that predicts classes for X.
- a `predict_proba(X)` method that predicts class probabilities for X.

# Metrics

We will have access to y_test (the ground truth) and y_pred (the predicted class) of each test sample.

From that we will be able to compute anyone of those metrics https://scikit-learn.org/stable/modules/classes.html#classification-metrics

# Proposed process

## When to fit

We could score our models 5 times over the fitting process (at 20%, 40%, 60%, 80% and 100%). That would allow us to visualise the evolution of our metrics when the model fits more data.

Otherwise, we can just score our models at the end of the fitting process.

## How to store

In each case, we would store y_test and y_pred so if we do multiple evaluations per model, we would have something like that:

```json
{
    "data_augmentation": {
        "...": "..."
    }
    "classifier": {
        "name": "...",
        "hyperparameters": {
            "...": "..."
        }
    }
    "predictions": [
        {
            "y_test": [1, 0, 3, 2, 3, 3, 1, 1, 0],
            "y_pred": [1, 1, 3, 3, 3, 2, 1, 1, 0]
        },
        {
            "y_test": [1, 0, 3, 2, 3, 3, 1, 1, 0],
            "y_pred": [1, 1, 3, 3, 3, 2, 1, 1, 0]
        },
        "..."
    ]
}
```

**Comment by MarcBresson** [comment]
@ArefBitar @Alcharyx @Pradnya1208 what do you think I should implement?

**Comment by MarcBresson** [comment]
The new way the features are generated is by using a python generator that would return the features of **only one** skeleton over its rolling window. i.e. for a video containing 2 skeletons on 35 frames (minus 5 for the rolling window to fill up), we would fit our model `2*30=60` times. 

Because we would not know the entire feature set in advance, the usage of PCA is harder and biaised. We would have to fit the PCA on a too small feature set.

@Pradnya1208 should we also test the usage of PCA? With the generator, it is still possible to grow one big feature set (just as you are doing) and feed it to the model to fit.

**Reply by Alcharyx** [reply]
As long as model inference is not too long PCA is probably not needed + the teacher said that PCA wasn't recommended in our case so I think we can drop it for now

**Comment by MarcBresson** [comment]
It appears that it cannot be done without a partial_fit function. Nevertheless, the generator will be used to generate one big feature set.
